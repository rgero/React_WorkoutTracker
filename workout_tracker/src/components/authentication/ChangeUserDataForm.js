import React, {useState} from 'react';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';


export const ChangeUserFormData = ({currentUser, onSubmit})=> 
{
    const [email, setEmail] = useState(currentUser ? currentUser.email : "");
    const [displayName, setName] = useState(currentUser ? currentUser.displayName : "");
    const [currentPassword, setCurrentPassword] = useState("");
    const [newPassword, setNewPassword] = useState("");
    const [newPasswordRetype, setRetypePassword] = useState("");
    const [errMsg, setErrorMessage] = useState("")

    const trySignUp = async (event) => {
        event.preventDefault();
        if (currentPassword === "")
        {
            setErrorMessage("You need to fill out your current password");
        }

        if ( email === "" || displayName === "")
        {
            setErrorMessage("You must have an email and a display name.");
            return;
        }
        if (newPassword !== newPasswordRetype)
        {
            setErrorMessage("Passwords do not match.");
            return;
        }
        setErrorMessage("");

        let changes = {};
        if (newPassword !== "")
        {
            changes.password = newPassword;
        }
        changes.displayName = displayName;
        changes.email = email;

        await onSubmit(currentUser.email, currentPassword, changes);
    }

    return(
        <Container fluid="md">
            <Row className="justify-content-md-center">
                <Col>
                    <h3>User Profile</h3>
                    <Form onSubmit={trySignUp}>
                        <Form.Group className="mb-3" controlId="formEmail">
                            <Form.Label>E-mail</Form.Label>
                            <Form.Control
                                aria-label="email"
                                type="email"
                                value={email}
                                onChange={e => setEmail(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="displayName">
                            <Form.Label>Display Name</Form.Label>
                            <Form.Control
                                aria-label="displayName"
                                type="text"
                                value={displayName}
                                onChange={e => setName(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Current Password</Form.Label>
                            <Form.Control
                                aria-label="currentPassword"
                                type="password"
                                value={currentPassword}
                                onChange={e => setCurrentPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Enter new password</Form.Label>
                            <Form.Control
                                aria-label="reentry"
                                type="password"
                                value={newPassword}
                                onChange={e => setNewPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Re-enter new password</Form.Label>
                            <Form.Control
                                aria-label="reentry"
                                type="password"
                                value={newPasswordRetype}
                                onChange={e => setRetypePassword(e.target.value)}
                            />
                        </Form.Group>
                        <Button variant="outline-secondary" type="submit">
                            Submit changes
                        </Button>
                        { errMsg ? 
                            (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {errMsg}</Alert>
                                </Container>
                            ) : null
                        }
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}