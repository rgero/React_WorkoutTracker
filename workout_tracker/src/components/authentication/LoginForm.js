import React, {useState} from 'react';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';


export const LoginForm = ({onSubmit})=> 
{
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [errMsg, setErrorMessage] = useState("");

    const trySignIn = async (event) => {
        event.preventDefault();
        if ( email === "")
        {
            setErrorMessage("Error: Missing e-mail");
            return;
        }

        if ( password === "")
        {
            setErrorMessage("Error: Missing password");
            return;
        }
        setErrorMessage("");
        await onSubmit({email, password});
    }

    return(
        <Container fluid="md">
            <Row>
                <Col xs lg="3">
                    <Form onSubmit={trySignIn}>
                        <Form.Group className="mb-3" controlId="formEmail">
                            <Form.Label>E-mail</Form.Label>
                            <Form.Control
                                aria-label="email"
                                type="email"
                                value={email}
                                onChange={e => setEmail(e.target.value)}
                            />
                        </Form.Group>

                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Password</Form.Label>
                            <Form.Control
                                aria-label="password"
                                type="password"
                                value={password}
                                onChange={e => setPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Button variant="outline-secondary" type="submit">
                            Log In
                        </Button>
                        { errMsg ? 
                            (
                                <Container className="pt-4">
                                    <Alert variant='danger'>{errMsg}</Alert>
                                </Container>
                            ) : null
                        }
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}