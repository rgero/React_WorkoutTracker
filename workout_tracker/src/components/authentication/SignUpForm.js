import React, {useState} from 'react';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Container from 'react-bootstrap/Container';

export const SignUpForm = ({errorMessage, onSubmit})=> 
{
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [passwordSecond, setRetypePassword] = useState("");
    const [errMsg, setErrorMessage] = useState(errorMessage ? errorMessage : "")

    const trySignUp = async (event) => {
        event.preventDefault();
        if (passwordSecond !== password)
        {
            setErrorMessage("Passwords do not match");
            return;
        }
        setErrorMessage("");
        await onSubmit({email, password});
    }

    return(
        <Container fluid="md">
            <Form onSubmit={trySignUp}>
                <InputGroup className="mb-3">
                    <InputGroup.Text>E-Mail</InputGroup.Text>
                    <Form.Control type="email" value={email} onChange={e => setEmail(e.target.value)}/>
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Password</InputGroup.Text>
                    <Form.Control type="password" value={password} onChange={e => setPassword(e.target.value)}/>
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Re-Enter Password</InputGroup.Text>
                    <Form.Control type="password" value={passwordSecond} onChange={e => setRetypePassword(e.target.value)}/>
                </InputGroup>
                <Button variant="outline-secondary" type="submit">
                    Sign In
                </Button>
                {errMsg ? (
                    <h4>{errMsg}</h4>
                ): null}
            </Form>
        </Container>
    )


}