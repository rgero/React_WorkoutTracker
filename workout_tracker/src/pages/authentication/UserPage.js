import React from "react";

import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';

export const UserPage = () => {
    const {state, tryLocalSignin} = React.useContext(AuthContext);

    React.useEffect(()=> {
        tryLocalSignin();
    }, [])

    return (
        <Container className="pt-3">
            <Row>
                <Col><h2>User Profile</h2></Col>
            </Row>
            <Row>
                <Col>
                    Display Name
                </Col>
                <Col>
                    {state.displayName}
                </Col>
            </Row>
            <Row>
                <Col>
                    E-Mail
                </Col>
                <Col>
                    {state.email}
                </Col>
            </Row>
        </Container>
    );
};

export default UserPage;