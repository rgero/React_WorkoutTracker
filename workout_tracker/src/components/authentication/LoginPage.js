import React, { useContext, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import { Navigate } from "react-router-dom";
import {Context as AuthContext} from '../../context/AuthContext';

import { LoginForm } from './LoginForm';


export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        tryLocalSignin();
    }, [])

    if (state.token)
    {
        clearErrorMessage();
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container fluid="md">
                <h1>Login</h1>
                <LoginForm
                    errorMessage={state.errorMessage}
                    onSubmit={({email, password})=> signIn({email, password})}
                />
            </Container>
        )
    }
}