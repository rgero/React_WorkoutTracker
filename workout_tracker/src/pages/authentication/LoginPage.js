import React, { useContext, useEffect } from 'react';
import { Navigate } from "react-router-dom";

import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';
import { LoginForm } from '../../components/authentication/LoginForm';

export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        const tryToSignIn = async () => {
            await tryLocalSignin();
            if (state.token)
            {
                clearErrorMessage();
            }
        }

        tryToSignIn();
    }, [])

    if (state.token)
    {
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container>
                <Row>
                    <Col md={8} className="d-none d-lg-block">
                        <img src="/images/dumbbells.jpg" width="100%" />
                    </Col>
                    <Col xs={10} md={4}>
                        <LoginForm
                            onSubmit={({email, password})=> signIn({email, password})}
                        />
                        { state.errorMessage ? (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {state.errorMessage}</Alert>
                                </Container>
                            ) : null
                        }
                    </Col>
                </Row>
            </Container>
        )
    }
}

export default LoginPage;