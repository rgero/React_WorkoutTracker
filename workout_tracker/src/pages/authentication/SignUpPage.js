import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';
import { SignUpForm } from '../../components/authentication/SignUpForm';

export const SignUpPage = ()=> 
{
    const navigate = useNavigate();
    const {state, signUp} = useContext(AuthContext);

    if (state.token)
    {
        navigate('/dashboard');
    } else 
    {
        return (
            <Container>
                <Row>
                    <Col xs={10} md={4}>
                        <SignUpForm onSubmit={({email, displayName, password})=> signUp({email, displayName, password})}/>
                    </Col>
                    <Col md={8} className="d-none d-lg-block">
                        <img src="/images/dumbbells.jpg" width="100%" />
                    </Col>
                </Row>

            </Container>
        )
    }
}

export default SignUpPage;