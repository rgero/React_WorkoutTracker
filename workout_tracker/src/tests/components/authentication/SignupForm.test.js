import TestRenderer from 'react-test-renderer';

import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import user from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import {SignUpForm} from '../../../components/authentication/SignUpForm';

describe("SignUpForm Rendering Tests", ()=> {
    test("No Data", ()=> {
        let element = new TestRenderer.create(
            <SignUpForm onSubmit={jest.fn()} />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })
})

describe("SignUpForm value tests", ()=> {
    let mockSubmit = jest.fn();

    let mockUserData = {
        email: "gibgibgib@gib.com",
        displayName: "Gib Gib",
        password: "NotARealPasswordObvs"
    }

    beforeEach(()=> {
        render(<SignUpForm onSubmit={mockSubmit} />)
    })

    test("Valid Submission", ()=> {
        act(()=> {
            const emailInput = screen.getByRole("textbox", {name: /email/i});
            const displayNameInput = screen.getByRole("textbox", {name: /displayname/i});
            const passwordInput = screen.getByLabelText(/password/i);
            const secondPassword = screen.getByLabelText(/reentry/i);

            user.click(emailInput);
            user.keyboard(mockUserData.email);

            user.click(displayNameInput);
            user.keyboard(mockUserData.displayName);
        
            user.click(passwordInput);
            user.keyboard(mockUserData.password);

            user.click(secondPassword);
            user.keyboard(mockUserData.password);
        
            const button = screen.getByRole("button");

            user.click(button);
        })

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith( mockUserData )
    })

    test("Missing email", ()=> {
        act(()=> {
            const displayNameInput = screen.getByRole("textbox", {name: /displayname/i});
            const passwordInput = screen.getByLabelText(/password/i);
            const secondPassword = screen.getByLabelText(/reentry/i);
        
            user.click(displayNameInput);
            user.keyboard(mockUserData.displayName);

            user.click(passwordInput);
            user.keyboard(mockUserData.password);

            user.click(secondPassword);
            user.keyboard(mockUserData.password);
        
            const button = screen.getByRole("button");

            user.click(button);
        })

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Please fill out the full form.");
        expect(errorField).toBeInTheDocument();
    })

    test("Missing first password", ()=> {
        act(()=> {
            const emailInput = screen.getByRole("textbox", {name: /email/i});
            const displayNameInput = screen.getByRole("textbox", {name: /displayname/i});
            const secondPassword = screen.getByLabelText(/reentry/i);

            user.click(emailInput);
            user.keyboard(mockUserData.email);

            user.click(displayNameInput);
            user.keyboard(mockUserData.displayName);

            user.click(secondPassword);
            user.keyboard(mockUserData.password);
        
            const button = screen.getByRole("button");

            user.click(button);
        })

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Please fill out the full form.");
        expect(errorField).toBeInTheDocument();
    })

    test("Missing second password", ()=> {
        act(()=> {
            const emailInput = screen.getByRole("textbox", {name: /email/i});
            const displayNameInput = screen.getByRole("textbox", {name: /displayname/i});
            const passwordInput = screen.getByLabelText(/password/i);

            user.click(emailInput);
            user.keyboard(mockUserData.email);

            user.click(displayNameInput);
            user.keyboard(mockUserData.displayName);
        
            user.click(passwordInput);
            user.keyboard(mockUserData.password);
        
            const button = screen.getByRole("button");

            user.click(button);
        })

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Please fill out the full form.");
        expect(errorField).toBeInTheDocument();
    })

    test("Missing display name", ()=> {
        act(()=> {
            const emailInput = screen.getByRole("textbox", {name: /email/i});
            const passwordInput = screen.getByLabelText(/password/i);
            const secondPassword = screen.getByLabelText(/reentry/i);

            user.click(emailInput);
            user.keyboard(mockUserData.email);
        
            user.click(passwordInput);
            user.keyboard(mockUserData.password);

            user.click(secondPassword);
            user.keyboard(mockUserData.password);
        
            const button = screen.getByRole("button");

            user.click(button);
        })

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Please fill out the full form.");
        expect(errorField).toBeInTheDocument();
    })

    test("Passwords won't match", ()=> {
        act(()=> {
            const emailInput = screen.getByRole("textbox", {name: /email/i});
            const displayNameInput = screen.getByRole("textbox", {name: /displayname/i});
            const passwordInput = screen.getByLabelText(/password/i);
            const secondPassword = screen.getByLabelText(/reentry/i);

            user.click(emailInput);
            user.keyboard(mockUserData.email);

            user.click(displayNameInput);
            user.keyboard(mockUserData.displayName);
        
            user.click(passwordInput);
            user.keyboard(mockUserData.password);

            let secondaryPassword = "Not the same password";
            user.click(secondPassword);
            user.keyboard(secondaryPassword);
        
            const button = screen.getByRole("button");
            
            user.click(button);
        })

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Passwords do not match.");
        expect(errorField).toBeInTheDocument();
    })
})