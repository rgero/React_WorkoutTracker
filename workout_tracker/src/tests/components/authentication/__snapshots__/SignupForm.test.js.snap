// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm Rendering Tests No Data 1`] = `
<div
  className="container-md"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col-lg-3 col"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formEmail"
          >
            E-mail
          </label>
          <input
            aria-label="email"
            className="form-control"
            id="formEmail"
            onChange={[Function]}
            type="email"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formPassword"
          >
            Password
          </label>
          <input
            aria-label="password"
            className="form-control"
            id="formPassword"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formPassword"
          >
            Re-enter password
          </label>
          <input
            aria-label="reentry"
            className="form-control"
            id="formPassword"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <button
          className="btn btn-outline-secondary"
          disabled={false}
          type="submit"
        >
          Sign Up
        </button>
      </form>
    </div>
  </div>
</div>
`;
